[DI] Deprecate Container::isFrozen and introduce isCompiled

ro0NL opened the pull request:
| Q | A |
| --- | --- |
| Branch? | "master" |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | yes |
| Tests pass? | yes |
| Fixed tickets | comma-separated list of tickets fixed by the PR, if any |
| License | MIT |
| Doc PR | reference to the documentation PR, if any |

This deprecates the concept of freezing a container, implied by `Container::isFrozen`. However, freezing happens due compilation (`Container::compile`). So having just `isCompiled` instead seems more intuitive, and plays along well with `ContainerBuilder`.

Before/After;
- `Container::isFrozen`
  - Checks if the parameter bag is frozen, but is deprecated in 3.2
  - In 4.0 this methods does not exists and can be replaced with `getParameterBag() instanceof FrozenParameterBag` _or_ `isCompiled()`. Depending on what you want (to clarify; the behavior is different when passing a frozen bag to the constructor)
- `Container::isCompiled`
  - Truly checks if `compile()` has ran, and is a new feature
- `ContainerBuilder::merge` etc.
  - Now uses `isCompiled` instead of `isFrozen`, ie. we allow for it till compilation regarding the state of the paramater bag

ro0NL committed:
[DI] Deprecate Container::isFrozen and introduce isCompiled

ro0NL commented:
In some ways it was actually compiling (resolving the parameters), as this is also done in a compiler pass. So we can also just go with `compile()` + `isCompiled()` on both sides. And deprecate the concept of freezing.

stof commented:
-1 for renaming compile to freeze. This method is not only about freezing. It is about performing all the compilation steps of the container building (everything done in compiler passes for instance).

ro0NL commented:
@stof different approach. WDYT?

fabpot commented:
I don't understand the problem you are trying to fix here. Can you elaborate? Deprecating something should be taken lightly, so we need to have a very good reason to rename something.

ro0NL commented:
It's confusing to mix&match `freeze/frozen` and `compile/compiled`. `isFrozen` is currently enabled due compilation. The parameter bag is frozen, the container compiled.

This exposes a minor code smell in `ContainerBuilder` where it's additionally tracking state (`ContainerBuilder::$compiled`) where this is actually the same as checking if it's frozen (ie. compiled). I guess the difference was unclear back then, and you get "better be safe then sorry" changes.

ro0NL commented on src/Symfony/Component/DependencyInjection/Container.php:
```diff
@@ -89,18 +90,38 @@ public function __construct(ParameterBagInterface $parameterBag = null)
-     * @return bool true if the container parameter bag are frozen, false otherwise
+     * @return bool
+     */
+    final public function isCompiled()
```
This is now the ought to use method. If you truely want to check for a frozen parameter bag use `getParameterBag() instanceof FrozenParameterBag` instead (hence the deprecation).

Besides that this method guarantees state of `compile()` as we dont depend on a frozen parameterbag for that anymore (which we can pass to the constructor ;-)).

nicolas-grekas replied:
Why making it final? That would be the first method in the class.

Ener-Getick commented on src/Symfony/Component/DependencyInjection/ContainerBuilder.php:
```diff
@@ -274,13 +272,13 @@ public function addClassResource(\ReflectionClass $class)
-        if ($this->isFrozen()) {
-            throw new BadMethodCallException('Cannot load from an extension on a frozen container.');
+        if ($this->isCompiled()) {
+            throw new BadMethodCallException('Cannot load from an extension on a conpiled container.');
```
`compiled`

ro0NL commented:
status: Ready

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Container.php:
```diff
@@ -89,18 +90,38 @@ public function __construct(ParameterBagInterface $parameterBag = null)
      */
     public function isFrozen()
     {
+        @trigger_error(sprintf('The %s method is deprecated as of 3.2 and will be removed in 4.0. Please use "%s() instanceof %s" instead.', __METHOD__, __CLASS__.'::getParameterBag', FrozenParameterBag::class), E_USER_DEPRECATED);
```
I'd suggest using `isCompiled` as replacement here.

ro0NL replied:
Guess so.. it's hard to guess what the user is using this method for right now.

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php:
```diff
@@ -140,7 +141,7 @@ public function dump(array $options = array())
         $code = $this->startClass($options['class'], $options['base_class'], $options['namespace']);
 
-        if ($this->container->isFrozen()) {
+        if ($this->container->getParameterBag() instanceof FrozenParameterBag) {
```
Would it work by using isCompiled? Looks more in line with the change to me.

ro0NL replied:
Havent figured it out yet, i think so.

[Tests](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/DependencyInjection/Tests/Dumper/PhpDumperTest.php#L123) somewhat imply this happens for compiled container builders.

nicolas-grekas commented:
I tend to be sympathetic to deprecating the "frozen" vocabulary on containers. It's been looking strange to me also and the difference is unclear with the "compile" concept. `isCompiled` looks better.
That may enhance DX, that would be the argument supporting this change.

ro0NL commented:
Updated the `PhpDumper` as well.. it makes sense, but can be unpredictable when you have `new Container(new FrozenParamaeterBag())`.

What do we expect when dumping this uncompiled?

ogizanagi commented:
@ro0NL : As we do not rely anymore on the `$this->parameterBag instanceof FrozenParameterBag`, I'd say it doesn't matter anymore... But...you'll probably have to check if the parameterBag is frozen in the `ContainerBuilder::merge()` method.

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Container.php:
```diff
@@ -89,18 +90,40 @@ public function __construct(ParameterBagInterface $parameterBag = null)
         $this->parameterBag->resolve();
 
-        $this->parameterBag = new FrozenParameterBag($this->parameterBag->all());
+        if (!$this->parameterBag instanceof FrozenParameterBag) {
```
This should be removed: changing the object instance of the bag can be seen as a wanted side effect of compiling. Not wrapping a FrozenParameterBag in a FrozenParameterBag looks like an unnecessary optimization to me also.

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Container.php:
```diff
@@ -89,18 +90,40 @@ public function __construct(ParameterBagInterface $parameterBag = null)
      */
     public function isFrozen()
     {
+        @trigger_error(sprintf('The %s method is deprecated as of 3.2 and will be removed in 4.0. Please use "%s" instead.', __METHOD__, __CLASS__.'::isCompiled'), E_USER_DEPRECATED);
```
`The %s() method is deprecated as of 3.2 and will be removed in 4.0. Use the isCompiled() method instead.`
Lets save repeating the (looong) namespace. And we don't "please" usually here :)

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php:
```diff
@@ -18,6 +18,7 @@
 use Symfony\Component\DependencyInjection\ContainerInterface;
 use Symfony\Component\DependencyInjection\Reference;
 use Symfony\Component\DependencyInjection\Parameter;
+use Symfony\Component\DependencyInjection\ParameterBag\FrozenParameterBag;
```
should be removed

ro0NL commented:
@ogizanagi merging is only allowed before compiling. If you passed a `FrozenParameterBag` at that point i tend to let it crash (from the paramereter bag side that is).

Not sure what's best for DX here.. crashing or ignoring silently...

@nicolas-grekas any thoughts?

ro0NL commented:
Oh and btw... just to be sure; the YAML+XML dumper really depend on a frozen bag right? Due escaping.

nicolas-grekas commented:
Rebase needed to take #19704 into account.

ro0NL commented:
Not sure.. but do we clean that up in master with this PR? Perhaps refactor the container to
`null === $this->parameterBag || instanceof`

Deprecations should be triggered from the dumped container, at least.

allflame commented:
@ro0NL as far as i can see isCompiled will return false for dumped container whether or not container was compiled beforehand

nicolas-grekas commented:
I really don't know if this has a chance to be accepted by the core team, but let's say yes, I think we should go one step further and try deprecating dumping non-compiled containers.
WDYT?

ro0NL commented:
@allflame `isCompiled` depends on `compile()`. The PHP dumper doesnt compile your container, it''s more or less designed to cover both cases, ie.

- dump a compiled container(builder)
- compile a dumped container(builder)

allflame commented:
@ro0NL I'm aware of that fact. Let me explain using your own statements:
Since `PhpDumper` has nothing to do with the compiled status, it shouldn't change it, right?
What I'm trying to say is that if you take your code, compile the container, you will get `isCompiled` resolving to true (correct), but if you dump it to the file and then create the container from the dumped file, you will get `isCompiled` resolving to false (incorrect), which is "strange" since you were dumping compiled container.
@nicolas-grekas tbh I'm not using the whole Symfony ecosystem so I cant think of implication this change can have (potential BC issues), but what I'm pretty sure of is that if you won't compile your container the "original" and "dumped" version won't necessarily be the same.

ro0NL commented:
@allflame sorry, you're right #19704 applies to `isCompiled` as well, i should fix that :+1: 

About BC, i truly consider this a bugfix and think the impact is minimal.

ro0NL commented:
@allflame i added the stub for `isCompiled`. Also rebased on master...

> I really don't know if this has a chance to be accepted by the core team, but let's say yes, I think we should go one step further and try deprecating dumping non-compiled containers.
WDYT?

Personally i find this out of scope for this PR. @fabpot WDYT by now?

ro0NL commented:
Rebased. Still all good :)

ro0NL commented:
I just realized i made mistake with the latest deprecation: https://github.com/symfony/symfony/commit/ed5b1d89685c5b808aaf1f119064554b84eccaa0#diff-f7b23d463cba27ac5e4cb677f2be7623R76

Im talking _uncompiled container_ here, looking at the class it _should_ have been _not-frozen container_. I really dont mind this tiny semantic. Then again, if we dont do the PR it might be worth updating that instead... sorry :)

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Container.php:
```diff
@@ -91,18 +92,37 @@ public function __construct(ParameterBagInterface $parameterBag = null)
+        $this->parameterBag->resolve();
         $this->parameterBag = new FrozenParameterBag($this->parameterBag->all());
+
+        $this->compiled = true;
```
I'd keep only this line a revert the previous changes in this method, that would prevent any potential BC-related side effect of the change.

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Container.php:
```diff
@@ -91,18 +92,37 @@ public function __construct(ParameterBagInterface $parameterBag = null)
      */
     public function isFrozen()
     {
+        @trigger_error(sprintf('The %s() method is deprecated as of 3.3 and will be removed in 4.0. Use the isCompiled() method instead.', __METHOD__), E_USER_DEPRECATED);
```
we use "since version 3.3" here usually I think

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/ContainerBuilder.php:
```diff
@@ -554,6 +554,10 @@ public function prependExtensionConfig($name, array $config)
     public function compile()
     {
+        if ($this->isCompiled()) {
+            return;
```
let's remove this change, it just opens a potential BC break to me

nicolas-grekas commented:
:+1: with suggested changes

ro0NL commented:
Updated, rebased & waiting for travis.

Cant wait for 4.0 guys :+1:

xabbuh commented:
@ro0NL Looks like tests fail.

ro0NL commented:
Hm.. looks fixture related, maybe something added in the meanwhile. Ill have a look tonigh;

Note to self;
- Symfony\Component\DependencyInjection\Tests\ContainerTest::testIsFrozen
- Symfony\Component\DependencyInjection\Tests\Dumper\PhpDumperTest::testClosureProxy

nicolas-grekas commented:
ping @ro0NL

ro0NL commented:
All good @nicolas-grekas

fabpot commented:
LGMT @ro0NL Can you rebase one last time? Thanks.

ro0NL commented:
@fabpot all good.

fabpot commented:
Thank you @ro0NL.

Taluu commented on src/Symfony/Component/DependencyInjection/ContainerBuilder.php:
```diff
@@ -493,13 +493,13 @@ public function getCompiler()
-        if ($this->isFrozen() && (isset($this->definitions[$id]) && !$this->definitions[$id]->isSynthetic())) {
+        if ($this->isCompiled() && (isset($this->definitions[$id]) && !$this->definitions[$id]->isSynthetic())) {
             // setting a synthetic service on a frozen container is alright
             throw new BadMethodCallException(sprintf('Setting service "%s" for an unknown or non-synthetic service definition on a frozen container is not allowed.', $id));
```
`sprintf('Setting service "%s" for an unknown or non-synthetic service definition on a compiled container is not allowed.', $id)`

basically, `s/frozen/compiled`, no ?

ro0NL replied:
:sob:

ro0NL replied:
Fixed with #22112
